### PR TITLE
References to milestones

### LINKED ISSUE #125: Issues with milestone migration and referencing
* Milestone references are not converted to links. Reference by number %nn or title using %"Milestone title" are possible and should be supported
* Deleted gitlab milestones introduce gaps in the sequence which would break number-based references if done naively.

One could introduce placeholder milestones as is done for issues, but I'm not sure that consistency in the numbering is really important. Instead one could save a mapping `GitHub milestone` -> `GitLab milestone` to be used when converting comments.

### PR DESCRIPTION
**Pull request by mdbenito**
This PR does again too many things (sorry...)

* It introduces new types representing Git(Lab|Hub) issues, MRs, and so on. It's kind of clunky for now but hopefully sets the stage for improvement.
* It removes some repetitious code and factors it out into `convertAssignees`, `convertMilestone`, `convertLabels`, which basically take those gitlab fields and make them into fields for requests to github
* It introduces a property `milestoneMap<number, Milestone>` for the GitHub helper which is necessary to make the substitutions of %nn and %"title" references to milestones by links to github milestones.
* Creates placeholders for deleted milestones in the same manner as for issues, to keep the numbering consistent (but with the `milestoneMap` it should be possible to avoid this)
* It moves all regex work to `convertIssuesAndComments()` (in particular, it removes generateUserProjectRegex()` from `utils.ts`. The logic is quite ugly and could use some improvement, but having it all in one place allows greater flexibility.

Closes #125

### CODE REVIEW

**Comment by mdbenito** [comment]
@spruce I've been working on a few changes allowing for consistent replacement of references to milestones, issues, merge requests and so on inside comments. In doing so, I've again introduced a lot of spurious changes, although I think they are for the best (basically added types, simplified some duplicate code, in particular some I had introduced, and a few more things which I detail above). This makes this to a PR which is difficult to review, sorry about that.

Anyway, this only addresses milestones for now. The main conceptual change is the introduction of the milestoneMap which allows lookup of references like `%"milestone"` in comments. We can use the same method  for merge requests later.

Besides this, there are a couple of new functions to simplify the code and a lot of noisy but (hopefully) harmless changes. Again, sorry for that.

PS: I'll start working on MR references when I find some more time.

File: src/index.ts [hunk 1]
```diff
@@ -198,67 +210,78 @@ async function migrate() {
...
-        console.error('Could not create milestone', milestone.title);
-        console.error(err);
-      }
+      await githubHelper
+        .createMilestone(milestone)
+        .then(created => {
+          let m = milestoneMap.get(milestone.iid);
+          if (m && m.number != created.number) {
+            throw new Error(
+              `Mismatch between milestone ${m.title} in map and created ${created.title}`
```
Comment (spruce) [inline src/index.ts:290]: Shouldn't this log the numbers and a title?

File: README.md [hunk 2]
```diff
@@ -64,7 +64,11 @@ GitLab's API [does not allow downloading of attachments](https://gitlab.com/gitl
 
 #### github.baseUrl
 
-Where is the github instance hosted? Default is the official `api.github.com` domain
```
[the remark this answers is not in the document]
Comment (mdbenito) [reply README.md:67]: For building links to milestones, labels, etc. in the comments

**Comment by mdbenito** [comment]
@spruce Latest conflicts solved

File: src/index.ts (hunk 1 above)
Comment (spruce) [reply src/index.ts]: sounds good

**Pull request merged**